style_editor.html

```diff
@@ -1432,12 +1432,32 @@
 
 // Create n textures of about 1MB each.
 function initGL() {
+  loadColorData();
+  // Clear existing tab links and tab bodies before populating
+  var tabLinksElement = FIND("TABLINKS");
+  var tabBodiesElement = FIND("TABBODIES");
+  tabLinksElement.innerHTML = "";
+  tabBodiesElement.innerHTML = "";
+
   AddTab("color", "Styles",effect_links.sort().join(""))
-  AddTab("rgb", "Colors",
-    qlinks.sort().join("") +
-    "<div class='custom-color'>" +
-    "<label for='COLOR'>Custom color </label>" +
-    "<input type='color' id='COLOR' value='#ff0000' class='color-picker' onclick='ClickColor()' /></div>");
+  if (colorsortState.get()) {
+    // Sort sortedRgbLinks based on the name:
+    sortedRgbLinks.sort((a, b) => a.name.localeCompare(b.name));
```

Comment:
ok. I think I've refreshed myself.
First, "qlinks" vs "sortedRgbLinks" is not very clear. How about if instead we use "sortedRgbLinks" for the chosen sorting, and then the arrays named like:
```
var byHueLinks = [];
var byNameLinks = [];
```
then populate them like this:
```
function AddColor(name, r, g, b) {
  colorNames[r+","+g+","+b] = name;
  byNameLinks.push({name, r, g, b});
  byHueLinks.push(mkcolorbutton(name, r, g, b));
  all_colors[name] = new RgbClass(r, g, b);
}
```
Then we can Create the empty tab in initGL() with just
```
AddTab("rgb", "Colors", ""); updateRgbTabContent();
```
Then populate the tab and update it either by name or by hue based off the settings checkbox like:
```
function SetTabContent(tab, data) {
  FIND(tab + "_tabcontent").innerHTML = data;
}

function updateRgbTabContent() {
 if (colorsortState.get()) {
    // Sort by name and map to HTML strings
    byNameLinks.sort((a, b) => a.name.localeCompare(b.name));
    sortedRgbLinks = byNameLinks.map(color => mkcolorbuttonByName(color.name, color.r, color.g, color.b));
  } else {
   // Sorting by hue is already good to go
    sortedRgbLinks = byHueLinks.sort();
  }
  // Populate rgb tab with chosen sort type
  SetTabContent("rgb", sortedRgbLinks.join("") +
    "<div class='custom-color'>" +
    "<label for='COLOR'>Custom color </label>" +
    "<input type='color' id='COLOR' value='#ff0000' class='color-picker' onclick='ClickColor()' /></div>");
}
```
with the maintained saved state driving it:
```
var colorsortState = new SavedState("colorsort", false, (on) => {
  updateRgbTabContent();
});
```

??

Reply:
This is much better, but there is no point in having "byHueLinks" and "byNameLinks", and it's also confusing that one of them has r,g,b,name and the other is an array of strings (with html elements in each string). Variables that have similar names ought to have similar content as well.

Since byHueLinks can easily be generated from byNameLinks, I think we should get rid of byHueLinks and rename 'byNameLinks' to 'colorData', or maybe simply 'colors'.

Ideally, updateRgbTabContent() would look something like this:

```cpp
function updateRgbTabContent() {
 if (colorsortState.get()) {
    colorData.sort((a, b) => a.name.localeCompare(b.name));
  } else {
    colorData.sort(sortByHue);
  }
  sortedRgbLinks = byNameLinks.map(color => mkcolorbuttonByName(color.name, color.r, color.g, color.b));

  // Populate rgb tab with chosen sort type
  SetTabContent("rgb", sortedRgbLinks.join("") +
    "<div class='custom-color'>" +
    "<label for='COLOR'>Custom color </label>" +
    "<input type='color' id='COLOR' value='#ff0000' class='color-picker' onclick='ClickColor()' /></div>");
}
```

But that means writing a new function called sortByHue which replaces my crude A: B: C: hack which allows string sorting to sort by hue.

If this is difficult, then we could just do it like this:

```cpp
function updateRgbTabContent() {
  colorData.sort((a, b) => a.name.localeCompare(b.name));
  sortedRgbLinks = byNameLinks.map(color => mkcolorbuttonByName(color.name, color.r, color.g, color.b));
  if (!colorsortState.get()) sortedRgbLinks.sort();

  // Populate rgb tab with chosen sort type
  SetTabContent("rgb", sortedRgbLinks.join("") +
    "<div class='custom-color'>" +
    "<label for='COLOR'>Custom color </label>" +
    "<input type='color' id='COLOR' value='#ff0000' class='color-picker' onclick='ClickColor()' /></div>");
}
```

Reply:
I had tried this, unfortunately ` sortedRgbLinks.sort();` does not sort by hue properly.
That creates a dynamic element like
```
<input type="button" style="border: 1px solid black;padding:8px;color:black;background:rgb(0,238,255)" class="btn" onclick="SetTo(&quot;BrainFreeze&quot;)" value="BrainFreeze">
```
Where the other way uses srt from the sortstring like
```
<input srt="A:0.044444444444444446" type="button" style="border: 1px solid black;padding:8px;color:black;background:rgb(255,68,0)" class="btn" onclick="SetTo(&quot;OrangeRed&quot;)" value="OrangeRed">
```

Reply:
Why don't they create the same thing?

Reply:
ok fresh coffee, fresh eyes.
Populate a single array in AddColor using the original logic that results in a proper sort by hue by just applying sort() later:
```
function AddColor(name, r, g, b) {
  colorNames[r+","+g+","+b] = name;
  colorData.push(mkcolorbutton(name, r, g, b));
  all_colors[name] = new RgbClass(r, g, b);
}
```
Remove mkcolorbuttonByName function as we'll do that later and simpler in updateRgbTabContent().

Then just get the name to sort by in a sortByName() function, or sort() the array by hue:
```
function SetTabContent(tab, data) {
  FIND(tab + "_tabcontent").innerHTML = data;
}

function sortByName() {
  // Extract name from HTML string
  colorData.sort((a, b) => {
    const nameA = a.match(/value='(.*?)'/)[1];
    const nameB = b.match(/value='(.*?)'/)[1];
    return nameA.localeCompare(nameB);
  });
  return colorData;
}

function updateRgbTabContent() {
  if (colorsortState.get()) {
    console.log("Sort colors by Name");
    sortedRgbLinks = sortByName();
  } else {
    console.log("Sort colors by Hue");
    sortedRgbLinks = colorData.sort();
  }

  SetTabContent("rgb", sortedRgbLinks.join("") +
    "<div class='custom-color'>" +
    "<label for='COLOR'>Custom color </label>" +
    "<input type='color' id='COLOR' value='#ff0000' class='color-picker' onclick='ClickColor()' /></div>");
}
```
Works well and it's quick to respond. Looks ok?
Note the already existing code for hue sort isn't exactly perfect once we get to snow, but that's another fix later.

![Screenshot 2023-12-26 at 4 50 43 PM](https://github.com/profezzorn/ProffieOS-StyleEditor/assets/53964195/739a1dbc-bb72-450f-b654-7ef81e647500)